lib/navigation/preprocess_imu.cpp

```diff
@@ -9,48 +15,158 @@ ImuPreprocessor::ImuPreprocessor()
 
 core::ImuData ImuPreprocessor::processData(const core::RawImuData raw_imu_data)
 {
-  /*
-  TODOLater: implement
-  basic plan:
-  - call detectOutliers. Return of that function is return of this function
-  - call checkReliable
-
-  */
-  return {0, 0, 0, 0};
+  core::ImuData clean_accelerometer_data;
+  clean_accelerometer_data = detectOutliers(raw_imu_data);
+  checkReliable();
+  return clean_accelerometer_data;
 }
 
 core::ImuData ImuPreprocessor::detectOutliers(const core::RawImuData imu_data)
 {
-  /*
-  TODOLater: implement
-  rough process:
-  - get q1, median, q3 of encoder array
-  - define upper & lower bounds as (-)1.5*inter-quatrile range
-  - if any datapoint outwith this, set
-  outlier_imus_[i] += 1
-  -set outlier points as median
-
-  -also has to be able to handle 1 unreliable sensor
-  -also figure out return type/ what we update and update
-  documentation as appropriate
-  */
-  return {0, 0, 0, 0};
+  core::ImuData clean_accelerometer_data;
+  // TODOLater : change magnitude for directions (hardware configuration?)
+  core::Float magnitude;
+  for (std::size_t i = 0; i < core::kNumImus; ++i) {
+    magnitude = 0;
+    for (std::size_t j = 0; j < 3; ++j) {
+      magnitude += std::pow(imu_data.at(i).at(j), 2);
+    }
+    clean_accelerometer_data.at(i) = std::sqrt(magnitude);
+  }
+  Quartiles quartiles     = getOutlierThresholds(clean_accelerometer_data);
+  core::Float iqr         = quartiles.q3 - quartiles.q1;
+  core::Float lower_bound = quartiles.median - 1.5 * iqr;
+  core::Float upper_bound = quartiles.median + 1.5 * iqr;
+  for (size_t i = 0; i < core::kNumImus; ++i) {
+    // converts outliers or unreliables to medians, updates number of consecutive outliers for each
+    // sensor
+    if (are_imus_reliable_.at(i) == false) {
+      clean_accelerometer_data.at(i) = quartiles.median;
+    } else if (clean_accelerometer_data.at(i) < lower_bound
+               || clean_accelerometer_data.at(i) > upper_bound) {
+      clean_accelerometer_data.at(i) = quartiles.median;
+      ++num_outliers_per_imu_.at(i);
+    } else {
+      num_outliers_per_imu_.at(i) = 0;
+    }
+  }
+  return clean_accelerometer_data;
 }
 
-void ImuPreprocessor::checkReliable(const core::ImuData &imu_data)
+Quartiles ImuPreprocessor::getOutlierThresholds(const core::ImuData &clean_accelerometer_data)
 {
+  const uint8_t num_reliable_accelerometers
+    = std::accumulate(are_imus_reliable_.begin(), are_imus_reliable_.end(), 0);
+
+  if (num_reliable_accelerometers == core::kNumImus) {
+    core::ImuData clean_accelerometer_data_copy;
+    std::copy(clean_accelerometer_data.begin(),
+              clean_accelerometer_data.end(),
+              clean_accelerometer_data_copy.begin());
+    std::sort(clean_accelerometer_data_copy.begin(), clean_accelerometer_data_copy.end());
+    const core::Float q1
```

Comment:
directly assign instead of making objects and then assigning the object to the struct member?